engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -1988,92 +1977,249 @@ public void storageMigration(final String vmUuid, final StoragePool destPool) {
     private void orchestrateStorageMigration(final String vmUuid, final StoragePool destPool) {
         final VMInstanceVO vm = _vmDao.findByUuid(vmUuid);
 
-        if (destPool == null) {
-            throw new CloudRuntimeException("Unable to migrate vm: missing destination storage pool");
-        }
+        preStorageMigrationStateCheck(destPool, vm);
 
         try {
-            stateTransitTo(vm, VirtualMachine.Event.StorageMigrationRequested, null);
-        } catch (final NoTransitionException e) {
-            s_logger.debug("Unable to migrate vm: " + e.toString());
-            throw new CloudRuntimeException("Unable to migrate vm: " + e.toString());
+            if(s_logger.isDebugEnabled()) {
+                s_logger.debug(
+                        String.format("Offline migration of %s vm %s with volumes",
+                                vm.getHypervisorType().toString(),
+                                vm.getInstanceName()));
+            }
+
+            migrateThroughHypervisorOrStorage(destPool, vm);
+
+        } catch (ConcurrentOperationException
+                | InsufficientCapacityException // possibly InsufficientVirtualNetworkCapacityException or InsufficientAddressCapacityException
+                | StorageUnavailableException e) {
+            String msg = String.format("Failed to migrate vm: %s" , vm.getUuid(), e);
+            s_logger.debug(msg);
+            throw new CloudRuntimeException(msg);
```

Comment:
what about re throwing the exception? So, we do not loose the initial stack trace.
I mean, you can send the original exception in the new `CloudRuntimeException`

Reply:
Good point, changed to "throw new CloudRuntimeException(msg, e);" and removed the exception from the debug message.